zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/usertask/processors/UserTaskCompleteProcessor.java

```diff
@@ -80,10 +84,29 @@ public void onFinalizeCommand(
     userTaskRecord.setVariables(command.getValue().getVariablesBuffer());
     userTaskRecord.setAction(command.getValue().getActionOrDefault(DEFAULT_ACTION));
 
+    // It's important to retrieve the user task record request metadata before appending the
+    // "COMPLETED" event, as it will be cleared by the "COMPLETED" event applier
+    final var recordRequestMetadata = userTaskState.getRecordRequestMetadata(userTaskKey);
     stateWriter.appendFollowUpEvent(userTaskKey, UserTaskIntent.COMPLETED, userTaskRecord);
     completeElementInstance(userTaskRecord);
-    responseWriter.writeEventOnCommand(
-        userTaskKey, UserTaskIntent.COMPLETED, userTaskRecord, command);
+
+    // If the record request metadata is present, it indicates that "complete" task listeners were
+    // configured, and that the normal flow of the "COMPLETE" command was interrupted by the
+    // "COMPLETE_TASK_LISTENER" command. In this case, we need to use the `requestId` and
+    // `requestStreamId` from the persisted metadata, which refers to the original "COMPLETE"
+    // command to write the response back.
+    recordRequestMetadata.ifPresentOrElse(
+        metadata ->
+            responseWriter.writeResponse(
+                userTaskKey,
+                UserTaskIntent.COMPLETED,
+                userTaskRecord,
+                ValueType.USER_TASK,
+                metadata.getRequestId(),
+                metadata.getRequestStreamId()),
+        () ->
+            responseWriter.writeEventOnCommand(
+                userTaskKey, UserTaskIntent.COMPLETED, userTaskRecord, command));
```

Comment:
🔧 It's probably never the case, but we could consider the case where the command doesn't have request metadata and it also wasn't stored in the state.

```suggestion
    if (command.requestId != null && command.requestStreamId() != null) {
      responseWriter.writeEventOnCommand(
          userTaskKey, UserTaskIntent.COMPLETED, userTaskRecord, command));
    } else {
      // If the record request metadata is present, it indicates that "complete" task listeners were
      // configured, and that the normal flow of the "COMPLETE" command was interrupted by the
      // "COMPLETE_TASK_LISTENER" command. In this case, we need to use the `requestId` and
      // `requestStreamId` from the persisted metadata, which refers to the original "COMPLETE"
      // command to write the response back.
      recordRequestMetadata.ifPresent(
        metadata ->
            responseWriter.writeResponse(
                userTaskKey,
                UserTaskIntent.COMPLETED,
                userTaskRecord,
                ValueType.USER_TASK,
                metadata.getRequestId(),
                metadata.getRequestStreamId()));
    }
```

Reply:
Thanks for the suggestion, @korthout! 🚀

I found that we can use the `command.hasRequestMetadata()` method to check if the current command contains request metadata. So, I’ve improved the previous implementation to only query request metadata from `UserTaskState` if it’s not present in the current command.

❓ Would it make sense to add `warn` logging when request metadata is missing in both: the current command and `UserTaskState`?
I noticed that logging isn’t widely used across command processors, but adding it here could improve visibility in such "unexpected" cases. I’d be glad to hear your thoughts on this!

Reply:
You're right that logging isn't widely used, but this is something [we'd like to improve](https://github.com/camunda/camunda/issues/16276). 

However, too many logs is also bad, so we need to consider whether they are useful. The way I think about it is: would it be helpful in this context when that case were to occur? Here, if we don't have any request info available, we can't respond. But that's expected I guess. Perhaps I'm overlooking something though. Curious to your thought 🙇

Reply:
🤔 Hmm… what I imagine is that having this log (e.g., `Request metadata was not present in the current command nor in user task state for user task with key {}. The client will not receive a response.`) could help clarify cases where the client didn’t receive a response - not due to a timeout issue (such as a long-running listener exceeded the Tasklist timeout value of 10-15 seconds (or whatever is configured)) but specifically because request metadata was unexpectedly missing.

Additionally, since this is a "_should not happen_" case, we normally wouldn’t expect to see it in the logs. So, its presence could serve as an indicator that something went wrong.

Or maybe I’m just overthinking it! 😄

Reply:
Anyway, we can always add the logging later if it turns out to be helpful!
For now, I’m merging it as is 😉